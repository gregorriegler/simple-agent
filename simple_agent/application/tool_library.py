from dataclasses import dataclass
from typing import Protocol, List, Optional

@dataclass
class ToolResult:
    feedback: str

    def __init__(self, feedback=""):
        self.feedback = feedback

    def __str__(self) : return self.feedback

@dataclass
class ContinueResult(ToolResult):
    def __init__(self, feedback=""):
        super().__init__(feedback)


@dataclass
class CompleteResult(ToolResult):
    pass

class ParsedTool(Protocol):
    arguments: str | None
    tool_instance: object

@dataclass
class MessageAndParsedTools:
    message: str
    tools: List[ParsedTool]


class Tool(Protocol):

    def get_usage_info(self) -> str:
        ...

    def execute(self, *args, **kwargs) -> ToolResult:
        ...


class ToolLibrary(Protocol):
    tools: List[Tool]

    def parse_tool(self, text: str) -> MessageAndParsedTools:
        ...

    def execute_parsed_tool(self, parsed_tool: ParsedTool) -> ToolResult:
        ...
