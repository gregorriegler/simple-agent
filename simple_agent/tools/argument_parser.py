import shlex


def create_lexer(text):
    lexer = shlex.shlex(text, posix=True)
    lexer.whitespace_split = True
    lexer.commenters = ''
    lexer.escape = ''
    return lexer


def split_arguments(text):
    lexer = create_lexer(text)
    return list(lexer)
