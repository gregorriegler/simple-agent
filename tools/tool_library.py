import re
import subprocess

from .ls_tool import LsTool
from .cat_tool import CatTool
from .create_file_tool import CreateFileTool
from .edit_file_tool import EditFileTool
from .subagent_tool import SubagentTool


class ParsedTool:
    """Represents a parsed tool with its name, arguments, and tool instance."""
    def __init__(self, name, arguments, tool_instance):
        self.name = name
        self.arguments = arguments
        self.tool_instance = tool_instance

    def __str__(self):
        args_str = f" {self.arguments}" if self.arguments else ""
        return f"Tool: {self.name}{args_str}"


class ToolLibrary:
    def __init__(self, message_claude = lambda messages, system_prompt: ""):
        self.message_claude = message_claude
        static_tools = self._create_static_tools()
        dynamic_tools = self._discover_dynamic_tools()


        self.tools = static_tools + dynamic_tools
        self._build_tool_dict()

    def _create_static_tools(self):
        """Create the core static tools that are always available."""
        return [
            LsTool(self.run_command),
            CatTool(self.run_command),
            CreateFileTool(self.run_command),
            EditFileTool(self.run_command),
            SubagentTool(self.run_command, self.message_claude)
        ]

    def _discover_dynamic_tools(self):
        """Discover and create dynamic tools from external sources."""
        return []

    def _build_tool_dict(self):
        self.tool_dict = {tool.name: tool for tool in self.tools}

    def get_tool_info(self, tool_name=None):
        """Get comprehensive information about tools"""
        if tool_name:
            tool = self.tool_dict.get(tool_name)
            if not tool:
                return f"Tool '{tool_name}' not found. Available tools: {', '.join(self.tool_dict.keys())}"

            if hasattr(tool, 'get_usage_info'):
                return tool.get_usage_info()
            else:
                return f"Tool: {tool.name}\nDescription: {getattr(tool, 'description', 'No description available')}"
        else:
            info_lines = ["Available Tools:"]
            for tool in self.tools:
                description = getattr(tool, 'description', 'No description available')
                info_lines.append(f"  {tool.name}: {description}")
            return "\n".join(info_lines)

    def parse_tool(self, text):
        pattern = r'^/([\w-]+)(?:\s+(.*))?$'

        if '\n' in text:
            return self._parse_multiline_command(text, pattern)

        return self._parse_single_line_command(text, pattern)

    def _parse_single_line_command(self, text, pattern):
        for line in text.splitlines():
            line = line.strip()
            match = re.match(pattern, line)
            if match:
                command, arguments = match.groups()
                tool = self.tool_dict.get(command)
                if not tool:
                    return None, f"Unknown command: {command}"
                else:
                    return ParsedTool(command, arguments, tool)
        return None

    def _parse_multiline_command(self, text, pattern):
        first_line = text.splitlines()[0].strip()
        match = re.match(pattern, first_line)
        if match:
            command, _ = match.groups()
            tool = self.tool_dict.get(command)
            if not tool:
                return None, f"Unknown command: {command}"
            else:
                if text.startswith(f"/{command} "):
                    arguments = text[len(f"/{command} "):]
                else:
                    arguments = None

                return ParsedTool(command, arguments, tool)
        return None

    def execute_parsed_tool(self, parsed_tool):
        args = parsed_tool.arguments.strip() if parsed_tool.arguments else None
        result = parsed_tool.tool_instance.execute(args)
        return result['output']

    @staticmethod
    def run_command(command, args=None, cwd=None):
        try:
            command_line = [command]
            if args:
                if isinstance(args, str):
                    args = [args]
                command_line += args

            result = subprocess.run(
                command_line,
                capture_output=True,
                text=True,
                encoding='utf-8',
                errors='replace',
                timeout=30,
                cwd=cwd
            )

            output = result.stdout.rstrip('\n')

            if result.stderr:
                if output:
                    output += f"\n";
                output += f"STDERR: {result.stderr}"

            return {
                'success': result.returncode == 0,
                'output': output,
                'returncode': result.returncode
            }
        except subprocess.TimeoutExpired:
            return {'success': False, 'output': 'Command timed out (30s limit)', 'returncode': -1}
        except Exception as e:
            return {'success': False, 'output': f'Error: {str(e)}', 'returncode': -1}
