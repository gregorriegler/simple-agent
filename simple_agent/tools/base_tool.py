import subprocess
from simple_agent.application.tool_library import ToolResult


class BaseTool:
    name = ''
    description = ''
    arguments = []

    def execute(self, *args, **kwargs) -> ToolResult:
        raise NotImplementedError("Subclasses must implement execute()")

    @staticmethod
    def run_command(command, args=None, cwd=None):
        try:
            command_line = [command]
            if args:
                if isinstance(args, str):
                    args = [args]
                command_line += args

            result = subprocess.run(
                command_line,
                stdin=subprocess.DEVNULL,
                stdout=subprocess.PIPE,
                stderr=subprocess.PIPE,
                text=True,
                encoding='utf-8',
                errors='replace',
                timeout=30,
                cwd=cwd
            )
            output = result.stdout.rstrip('\n')
            if result.stderr:
                if output:
                    output += f"\n";
                output += f"STDERR: {result.stderr}"
            return {
                'output': output
            }
        except subprocess.TimeoutExpired:
            return {'output': 'Command timed out (30s limit)'}
        except Exception as e:
            return {'output': f'Error: {str(e)}'}

    def get_usage_info(self):
        if hasattr(self, '_custom_usage_info'):
            return self._custom_usage_info()

        return self._generate_usage_info_from_metadata()

    def _generate_usage_info_from_metadata(self):
        lines = [f"Tool: {self.name}"]

        if self.description:
            lines.append(f"Description: {self.description}")

        if self.arguments:
            lines.append("")
            lines.append("Arguments:")
            for arg in self.arguments:
                required_str = " (required)" if arg.get('required', False) else " (optional)"
                type_str = f" - {arg['name']}: {arg.get('type', 'string')}{required_str}"
                if 'description' in arg:
                    type_str += f" - {arg['description']}"
                lines.append(type_str)

            # Generate syntax
            lines.append("")
            required_args = [f"<{arg['name']}>" for arg in self.arguments if arg.get('required', False)]
            optional_args = [f"[{arg['name']}]" for arg in self.arguments if not arg.get('required', False)]
            all_args = required_args + optional_args
            syntax = f"üõ†Ô∏è {self.name}"
            if all_args:
                syntax += " " + " ".join(all_args)
            lines.append(f"Usage: {syntax}")

        if hasattr(self, 'examples') and self.examples:
            lines.append("")
            lines.append("Examples:")
            for example in self.examples:
                lines.append(f"{example}")

        return "\n".join(lines)
