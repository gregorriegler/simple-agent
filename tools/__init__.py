from .base_tool import BaseTool
from .tool_library import ToolLibrary, ParsedTool

__all__ = [
    'BaseTool',
    'ToolLibrary',
    'ParsedTool'
]
