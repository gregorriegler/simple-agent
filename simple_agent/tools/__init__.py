from .base_tool import BaseTool
from .all_tools import AllTools, ParsedTool

__all__ = [
    'BaseTool',
    'AllTools',
    'ParsedTool'
]
