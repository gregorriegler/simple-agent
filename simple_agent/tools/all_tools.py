import re
from typing import Callable, Any

from simple_agent.application.agent import Agent
from simple_agent.application.event_bus_protocol import EventBus
from simple_agent.application.input import Input
from simple_agent.application.llm import LLM
from simple_agent.application.session_storage import NoOpSessionStorage
from simple_agent.application.tool_library import ToolLibrary, MessageAndParsedTools, ParsedTool, Tool
from simple_agent.system_prompt_generator import generate_system_prompt
from .bash_tool import BashTool
from .cat_tool import CatTool
from .complete_task_tool import CompleteTaskTool
from .create_file_tool import CreateFileTool
from .edit_file_tool import EditFileTool
from .ls_tool import LsTool
from .subagent_tool import SubagentTool
from .write_todos_tool import WriteTodosTool


class AllTools(ToolLibrary):
    def __init__(
        self,
        llm: LLM | None = None,
        indent_level=0,
        agent_id: str = "Agent",
        event_bus: EventBus | None = None,
        user_input=None,
        create_subagent_display: Callable[[str, int], Any] | None = None,
        create_subagent_input: Callable[[int], Input] | None = None
    ):
        from simple_agent.application.event_bus import SimpleEventBus

        self.llm: LLM = llm if llm is not None else (lambda system_prompt, messages: '')
        self.indent_level = indent_level
        self.agent_id = agent_id
        self.event_bus: EventBus = event_bus if event_bus is not None else SimpleEventBus()
        self.user_input = user_input

        assert create_subagent_display is not None, "create_subagent_display must be provided"
        assert create_subagent_input is not None, "create_subagent_input must be provided"

        self.create_subagent_display = create_subagent_display
        self.create_subagent_input = create_subagent_input

        static_tools = self._create_static_tools()
        dynamic_tools = self._discover_dynamic_tools()
        self.tools: list[Tool] = static_tools + dynamic_tools + [self._create_subagent_tool()]
        self.tool_dict = {getattr(tool, 'name', ''): tool for tool in self.tools}

    def _create_static_tools(self):
        return [
            WriteTodosTool(self.agent_id),
            LsTool(),
            CatTool(),
            CreateFileTool(),
            EditFileTool(),
            #            PatchFileTool(),
            #            RememberTool(),
            #           RecallTool(),
            CompleteTaskTool(),
            BashTool()
        ]

    def _discover_dynamic_tools(self):
        return []

    def _create_subagent_tool(self):
        return SubagentTool(
            self._create_default_agent,
            self.create_subagent_display,
            self.indent_level,
            self.agent_id,
            self.create_subagent_input
        )

    def _create_default_agent(
        self,
        agent_id,
        user_input,
        session_storage=NoOpSessionStorage()
    ) -> Agent:
        subagent_tools = AllTools(
            self.llm,
            self.indent_level + 1,
            agent_id,
            self.event_bus,
            self.user_input,
            self.create_subagent_display,
            self.create_subagent_input
        )
        system_prompt = self._build_system_prompt(subagent_tools)
        return Agent(
            agent_id,
            system_prompt,
            subagent_tools,
            self.llm,
            user_input,
            self.event_bus,
            session_storage
        )

    def _build_system_prompt(self, subagent_tools) -> Callable[[Any], str]:
        return lambda tool_library: generate_system_prompt(subagent_tools)

    def parse_message_and_tools(self, text) -> MessageAndParsedTools:
        pattern = r'^🛠️ ([\w-]+)(?:\s+(.*))?'
        end_marker = r'^🛠️🔚'
        lines = text.splitlines(keepends=True)
        parsed_tools = []
        message = ""
        first_tool_index = None

        i = 0
        while i < len(lines):
            match = re.match(pattern, lines[i], re.DOTALL)
            if match:
                if first_tool_index is None:
                    first_tool_index = i
                    message = ''.join(lines[:i]).rstrip()

                command, same_line_args = match.groups()
                tool = self.tool_dict.get(command)
                if not tool:
                    return MessageAndParsedTools(message=text, tools=[])

                all_arg_lines = []
                if same_line_args:
                    all_arg_lines.append(same_line_args)

                i += 1
                while i < len(lines) and not re.match(r'^🛠️ ', lines[i]) and not re.match(end_marker, lines[i]):
                    all_arg_lines.append(lines[i])
                    i += 1

                if i < len(lines) and re.match(end_marker, lines[i]):
                    i += 1

                arguments = ''.join(all_arg_lines).rstrip()
                parsed_tools.append(ParsedTool(command, arguments, tool))
            else:
                i += 1

        if parsed_tools:
            return MessageAndParsedTools(message=message, tools=parsed_tools)
        return MessageAndParsedTools(message=text, tools=[])

    def execute_parsed_tool(self, parsed_tool):
        args = parsed_tool.arguments if parsed_tool.arguments else None
        result = parsed_tool.tool_instance.execute(args)
        return result
