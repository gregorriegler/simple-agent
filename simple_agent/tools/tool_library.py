import re

from simple_agent.application.agent import Agent
from simple_agent.application.display import Display
from simple_agent.application.io import IO
from simple_agent.application.llm import LLM
from simple_agent.infrastructure.stdio import StdIO
from simple_agent.infrastructure.textual_display import TextualDisplay
from simple_agent.infrastructure.console_display import ConsoleDisplay
from simple_agent.system_prompt_generator import generate_system_prompt
from simple_agent.application.session_storage import NoOpSessionStorage
from . import subagent_tool
from .bash_tool import BashTool
from .cat_tool import CatTool
from .complete_task_tool import CompleteTaskTool
from .create_file_tool import CreateFileTool
from .edit_file_tool import EditFileTool
from .ls_tool import LsTool
from .subagent_tool import SubagentTool
from .write_todos_tool import WriteTodosTool


class SubagentConsoleDisplay(ConsoleDisplay):
    def __init__(self, indent_level=1, io: IO | None = None):
        super().__init__(indent_level, "Subagent", io)

    def exit(self):
        pass


class SubagentTextualDisplay(TextualDisplay):
    def __init__(self, indent_level=1, io: IO | None = None):
        super().__init__(indent_level, "Subagent", io)

    def exit(self):
        pass


class ParsedTool:
    def __init__(self, name, arguments, tool_instance):
        self.name = name
        self.arguments = arguments
        self.tool_instance = tool_instance

    def __str__(self):
        if self.arguments:
            return f"🛠️ {self.name} {self.arguments}"
        return f"🛠️ {self.name}"


class AllTools:
    def __init__(
        self,
        llm: LLM | None = None,
        indent_level=0,
        io: IO | None = None,
        agent_id: str = "Agent",
        event_bus=None,
        display_event_handler=None
    ):
        if llm is None:
            llm = lambda system_prompt, messages: ''
        self.llm: LLM = llm
        self.indent_level = indent_level
        self.io = io or StdIO()
        self.agent_id = agent_id
        self.event_bus = event_bus
        self.display_event_handler = display_event_handler

        static_tools = self._create_static_tools()
        dynamic_tools = self._discover_dynamic_tools()
        self.tools = static_tools + dynamic_tools + [self._create_subagent_tool()]
        self.tool_dict = {tool.name: tool for tool in self.tools}

    def _create_static_tools(self):
        return [
            WriteTodosTool(self.agent_id),
            LsTool(),
            CatTool(),
            CreateFileTool(),
            EditFileTool(),
            #            PatchFileTool(),
            #            RememberTool(),
            #           RecallTool(),
            CompleteTaskTool(),
            BashTool()
        ]

    def _discover_dynamic_tools(self):
        return []

    def _create_subagent_tool(self):
        subagent_display = self._create_subagent_display()
        return SubagentTool(
            self._create_subagent,
            subagent_display,
            self.io,
            self.display_event_handler,
            self.agent_id,
            self.indent_level
        )

    def _create_subagent_display(self) -> Display:
        if self.display_event_handler:
            parent_display = self.display_event_handler.displays.get("Agent")
            if isinstance(parent_display, TextualDisplay):
                return SubagentTextualDisplay(self.indent_level + 1, self.io)
        return SubagentConsoleDisplay(self.indent_level + 1, self.io)

    def _create_subagent(
        self,
        parent_agent_id,
        subagent_counter,
        user_input,
        display_event_handler
    ):
        agent_id = f"{parent_agent_id}/Subagent{subagent_counter}"
        subagent_tools = AllTools(
            self.llm,
            self.indent_level + 1,
            self.io,
            agent_id,
            self.event_bus,
            display_event_handler
        )
        system_prompt = self._build_system_prompt(subagent_tools)
        session_storage = NoOpSessionStorage()
        return Agent(
            agent_id,
            system_prompt,
            subagent_tools,
            self.llm,
            user_input,
            self.event_bus,
            session_storage
        )

    def _build_system_prompt(self, subagent_tools):
        return lambda tool_library: generate_system_prompt(subagent_tools)

    def get_tool_info(self, tool_name=None):
        if tool_name:
            tool = self.tool_dict.get(tool_name)
            if not tool:
                return f"Tool '{tool_name}' not found. Available tools: {', '.join(self.tool_dict.keys())}"
            if hasattr(tool, 'get_usage_info'):
                return tool.get_usage_info()
            else:
                return f"Tool: {tool.name}\nDescription: {getattr(tool, 'description', 'No description available')}"
        else:
            info_lines = ["Available Tools:"]
            for tool in self.tools:
                description = getattr(tool, 'description', 'No description available')
                info_lines.append(f"  {tool.name}: {description}")
            return "\n".join(info_lines)

    def parse_tool(self, text):
        pattern = r'^🛠️ ([\w-]+)(?:\s+(.*))?'
        lines = text.splitlines(keepends=True)
        for i, line in enumerate(lines):
            match = re.match(pattern, line, re.DOTALL)
            if match:
                command, same_line_args = match.groups()
                tool = self.tool_dict.get(command)
                if not tool:
                    return None
                all_arg_lines = []
                if same_line_args:
                    all_arg_lines.append(same_line_args)
                for j in range(i + 1, len(lines)):
                    if re.match(r'^🛠️ ', lines[j]):
                        break
                    all_arg_lines.append(lines[j])
                arguments = ''.join(all_arg_lines)
                return ParsedTool(command, arguments, tool)
        return None

    def execute_parsed_tool(self, parsed_tool):
        args = parsed_tool.arguments if parsed_tool.arguments else None
        result = parsed_tool.tool_instance.execute(args)
        return result
