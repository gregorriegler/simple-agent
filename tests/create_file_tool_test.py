from approvaltests import Options, verify

from simple_agent.tools.tool_library import ToolLibrary
from .test_helpers import all_scrubbers, temp_directory

library = ToolLibrary()


def test_create_tool_single_character_name(tmp_path):
    verify_create_tool(library, "🛠️ create-file a", "a", tmp_path=tmp_path)


def test_create_tool_simple_name_with_extension(tmp_path):
    verify_create_tool(library, "🛠️ create-file test.txt", "test.txt", tmp_path=tmp_path)


def test_create_tool_single_character_content(tmp_path):
    verify_create_tool(library, "🛠️ create-file test.txt a", "test.txt", tmp_path=tmp_path)


def test_create_tool_simple_text_content(tmp_path):
    verify_create_tool(library, "🛠️ create-file readme.txt Hello World", "readme.txt", tmp_path=tmp_path)


def test_create_tool_newline_content(tmp_path):
    verify_create_tool(library, '🛠️ create-file multi.txt "Line 1\\nLine 2"', "multi.txt", tmp_path=tmp_path)


def test_create_tool_json_content(tmp_path):
    verify_create_tool(library, '🛠️ create-file config.json {"name": "test"}', "config.json", tmp_path=tmp_path)


def test_create_tool_filename_with_spaces(tmp_path):
    verify_create_tool(library, '🛠️ create-file "notes folder/note file.txt" "First line\\nSecond line"',
                      "notes folder/note file.txt", tmp_path=tmp_path)


def test_create_tool_content_with_leading_and_trailing_spaces(tmp_path):
    verify_create_tool(library, '🛠️ create-file padded.txt "  spaced content  "', "padded.txt", tmp_path=tmp_path)


def test_create_tool_explicit_empty_content(tmp_path):
    verify_create_tool(library, '🛠️ create-file empty.txt ""', "empty.txt", tmp_path=tmp_path)


def test_create_file_in_nonexistent_directory(tmp_path):
    verify_create_tool(library, f'🛠️ create-file src/utils/helper.py "# Helper module"', "src/utils/helper.py",
                       tmp_path=tmp_path)


def test_create_file_already_exists(tmp_path):
    with temp_directory(tmp_path):
        tool = library.parse_tool("🛠️ create-file existing.txt")
        library.execute_parsed_tool(tool)

        tool = library.parse_tool("🛠️ create-file existing.txt")
        result = library.execute_parsed_tool(tool)
        assert 'already exists' in result.feedback.lower() or 'exists' in result.feedback.lower()


def test_create_tool_on_second_line(tmp_path):
    verify_create_tool(library, "let me create a file\n🛠️ create-file a", "a", tmp_path=tmp_path)


def test_create_tool_on_second_line_with_multiline_content(tmp_path):
    verify_create_tool(library, "let me create a file\n🛠️ create-file test.txt Line1\nLine2", "test.txt", tmp_path=tmp_path)


def test_create_tool_stops_at_next_command(tmp_path):
    verify_create_tool(library, "🛠️ create-file test.txt Line1\nLine2\n🛠️ ls", "test.txt", tmp_path=tmp_path)


def test_create_tool_content_with_single_quotes(tmp_path):
    verify_create_tool(library, '🛠️ create-file test.txt \'He said "Hello"\'', "test.txt", tmp_path=tmp_path)


def test_create_tool_json_with_nested_quotes(tmp_path):
    verify_create_tool(library, '🛠️ create-file config.json \'{"name": "John", "message": "Hello \\"World\\""}\'', "config.json", tmp_path=tmp_path)


def test_create_tool_content_with_backslashes(tmp_path):
    verify_create_tool(library, '🛠️ create-file paths.txt "Path: C:\\\\Users\\\\test"', "paths.txt", tmp_path=tmp_path)


def test_create_tool_content_with_tabs(tmp_path):
    verify_create_tool(library, '🛠️ create-file table.txt "Column1\\tColumn2\\tColumn3"', "table.txt", tmp_path=tmp_path)


def test_create_tool_content_with_mixed_special_characters(tmp_path):
    verify_create_tool(library, '🛠️ create-file special.txt "Special: @#$%^&*()_+-={}[]|\\\\:;\'<>?,./"', "special.txt", tmp_path=tmp_path)


def test_create_tool_content_with_unicode_characters(tmp_path):
    verify_create_tool(library, '🛠️ create-file unicode.txt "Unicode: 🚀 café naïve résumé"', "unicode.txt", tmp_path=tmp_path)


def test_create_tool_content_with_single_quotes_inside_double_quotes(tmp_path):
    verify_create_tool(library, '🛠️ create-file test.txt "It\'s a test"', "test.txt", tmp_path=tmp_path)


def test_create_tool_multi_line_content(tmp_path):
    verify_create_tool(library,
"""
🛠️ create-file a First Line
Second Line
Third Line
""", "a", tmp_path=tmp_path)


def verify_create_tool(library, command, expected_filename, tmp_path):
    with temp_directory(tmp_path):
        tool = library.parse_tool(command)
        result = library.execute_parsed_tool(tool)

        with open(expected_filename, "r", encoding='utf-8') as f:
            actual_content = f.read()
            file_info = f"File created: {expected_filename}\nFile content:\n--- FILE CONTENT START ---\n{actual_content}\n--- FILE CONTENT END ---"

        verify(
            f"Command:\n{command}\n\nResult:\n{result}\n\n{file_info}",
            options=Options().with_scrubber(all_scrubbers())
        )
