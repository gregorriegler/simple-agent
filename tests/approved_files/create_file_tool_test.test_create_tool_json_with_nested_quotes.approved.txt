Command:
üõ†Ô∏è create-file config.json '{"name": "John", "message": "Hello \"World\""}'

Result:
Created file: config.json with content

File created: config.json
File content:
--- FILE CONTENT START ---
{"name": "John", "message": "Hello \"World\""}
--- FILE CONTENT END ---
